drivers/gpu/drm/sun4i/sun4i_hdmi_i2c_drv.c

```diff
@@ -0,0 +1,514 @@
+				  SUN4I_HDMI_DDC_OFFSET, 6, 6),
+	.field_ddc_sck_en	= REG_FIELD(SUN6I_HDMI_DDC_CTRL_REG +
+				  SUN4I_HDMI_DDC_OFFSET, 4, 4),
+
+	.ddc_fifo_reg		= SUN6I_HDMI_DDC_FIFO_DATA_REG +
+				  SUN4I_HDMI_DDC_OFFSET,
+	.ddc_fifo_thres_incl	= true,
+};
+
+static const struct regmap_config sun4i_hdmi_i2c_legacy_regmap_config = {
+	.reg_bits	= 32,
+	.val_bits	= 32,
+	.reg_stride	= 4,
+	.max_register	= 0x580,
+};
+
+static int fifo_transfer(struct sun4i_hdmi_i2c_drv *drv, u8 *buf, int len, bool read)
+{
+	/*
+	 * 1 byte takes 9 clock cycles (8 bits + 1 ACK) = 90 us for 100 kHz
+	 * clock. As clock rate is fixed, just round it up to 100 us.
+	 */
+	const unsigned long byte_time_ns = 100;
+	const u32 mask = SUN4I_HDMI_DDC_INT_STATUS_ERROR_MASK |
+			 SUN4I_HDMI_DDC_INT_STATUS_FIFO_REQUEST |
+			 SUN4I_HDMI_DDC_INT_STATUS_TRANSFER_COMPLETE;
+	u32 reg;
+	/*
+	 * If threshold is inclusive, then the FIFO may only have
+	 * RX_THRESHOLD number of bytes, instead of RX_THRESHOLD + 1.
+	 */
+	int read_len = RX_THRESHOLD +
+		(drv->variant->ddc_fifo_thres_incl ? 0 : 1);
+
+	/*
+	 * Limit transfer length by FIFO threshold or FIFO size.
+	 * For TX the threshold is for an empty FIFO.
+	 */
+	len = min_t(int, len, read ? read_len : SUN4I_HDMI_DDC_FIFO_SIZE);
+
+	/* Wait until error, FIFO request bit set or transfer complete */
+	if (regmap_field_read_poll_timeout(drv->field_ddc_int_status, reg,
+					   reg & mask, len * byte_time_ns,
+					   100000))
+		return -ETIMEDOUT;
+
+	if (reg & SUN4I_HDMI_DDC_INT_STATUS_ERROR_MASK)
+		return -EIO;
+
+	if (read)
+		readsb(drv->base + drv->variant->ddc_fifo_reg, buf, len);
+	else
+		writesb(drv->base + drv->variant->ddc_fifo_reg, buf, len);
+
+	/* Clear FIFO request bit by forcing a write to that bit */
+	regmap_field_force_write(drv->field_ddc_int_status,
+				 SUN4I_HDMI_DDC_INT_STATUS_FIFO_REQUEST);
+
+	return len;
+}
+
+static int xfer_msg(struct sun4i_hdmi_i2c_drv *drv, struct i2c_msg *msg)
+{
+	int i, len;
+	u32 reg;
+
+	/* Set FIFO direction */
+	if (drv->variant->ddc_fifo_has_dir) {
+		reg = readl(drv->base + SUN4I_HDMI_DDC_CTRL_REG);
+		reg &= ~SUN4I_HDMI_DDC_CTRL_FIFO_DIR_MASK;
+		reg |= (msg->flags & I2C_M_RD) ?
+		       SUN4I_HDMI_DDC_CTRL_FIFO_DIR_READ :
+		       SUN4I_HDMI_DDC_CTRL_FIFO_DIR_WRITE;
+		writel(reg, drv->base + SUN4I_HDMI_DDC_CTRL_REG);
+	}
+
+	/* Clear address register (not cleared by soft reset) */
+	regmap_field_write(drv->field_ddc_addr_reg, 0);
+
+	/* Set I2C address */
+	regmap_field_write(drv->field_ddc_slave_addr, msg->addr);
+
+	/*
+	 * Set FIFO RX/TX thresholds and clear FIFO
+	 *
+	 * If threshold is inclusive, we can set the TX threshold to
+	 * 0 instead of 1.
+	 */
+	regmap_field_write(drv->field_ddc_fifo_tx_thres,
+			   drv->variant->ddc_fifo_thres_incl ? 0 : 1);
+	regmap_field_write(drv->field_ddc_fifo_rx_thres, RX_THRESHOLD);
+	regmap_field_write(drv->field_ddc_fifo_clear, 1);
+	if (regmap_field_read_poll_timeout(drv->field_ddc_fifo_clear,
+					   reg, !reg, 100, 2000))
```

Comment:
introduce defines that describe what 100 and 2000 mean?
seconds? milisecond? usec?